modin/pandas/groupby.py

```diff
@@ -295,17 +304,130 @@ def apply(self, func, *args, **kwargs):
     def dtypes(self):
         if self._axis == 1:
             raise ValueError("Cannot call dtypes on groupby with axis=1")
+
+        def dtypes_getter(grp):
+            try:
+                return grp.dtypes
+            except AttributeError:
+                return grp.dtype
+
         if not self._as_index:
-            return self.apply(lambda df: df.dtypes)
+            return self.apply(dtypes_getter)
         else:
-            return self._apply_agg_function(lambda df: df.dtypes)
+            return self._apply_agg_function(dtypes_getter)
 
     def first(self, **kwargs):
         return self._default_to_pandas(lambda df: df.first(**kwargs))
 
     def backfill(self, limit=None):
         return self.bfill(limit)
 
+    _internal_by_cache = no_default
+
+    # TODO: since python 3.9:
+    # @cached_property
+    @property
+    def _internal_by(self):
+        """
+        Get only those components of 'by' that are column labels of the source frame.
+
+        Returns
+        -------
+        list of labels
+        """
+        if self._internal_by_cache is not no_default:
+            return self._internal_by_cache
+
+        internal_by = []
+        if self._drop:
+            for by in self._by if is_list_like(self._by) else [self._by]:
+                if isinstance(by, str):
+                    internal_by.append(by)
+                elif isinstance(by, BaseQueryCompiler):
+                    internal_by.extend(by.columns)
+            internal_by = self._df.columns.intersection(internal_by).tolist()
+
+        self._internal_by_cache = internal_by
+        return internal_by
+
+    _non_by_selection_cache = no_default
+
+    # TODO: since python 3.9:
+    # @cached_property
+    @property
+    def _non_by_selection(self):
+        """
+        Get selection filtered from the `by` columns in it.
+
+        Returns
+        -------
+        same type as ``self._selection``
+            ``self._selection`` without `by` columns in it.
+        """
+        if self._non_by_selection_cache is not no_default:
+            return self._non_by_selection_cache
+        selection = self._selection
+        if selection is not None:
+            internal_by = self._internal_by
+            selection = [col for col in self._selection_list if col not in internal_by]
+            if self.ndim == 1:
+                assert (
+                    len(selection) == 1 or len(selection) == 0
+                ), f"Single-dimensional object has non-single dimensional selection: {selection}."
+                selection = selection[0] if len(selection) == 1 else []
```

Comment:
Does this intentionally rewrite `selection` that is computed above?

Reply:
it's not rewriting, it's just post-processing the computed result for 1-dim objects